crates/astria-composer/src/searcher/executor.rs

```diff
@@ -164,28 +183,46 @@ impl Executor {
                             break Err(e).wrap_err("failed submitting bundle to sequencer");
                         }
                     }
+                    // reset timer for preempt bundle if it doesn't fill up within the sequencer block time
+                    block_timer.reset_immediately();
                 }
 
-                // receive new bundle for processing
-                Some(bundle) = self.new_bundles.recv(), if submission_fut.is_terminated() => {
+                // submit the next available bundle
+                Some(bundle) = bundler.get_next(), if submission_fut.is_terminated() => {
+                    debug!(
+                        bundle_size=?bundle.curr_size,
+                        "submitting bundle to sequencer"
+                    );
+
                     // TODO(https://github.com/astriaorg/astria/issues/476): Attach the hash of the
                     // bundle to the span. Linked GH issue is for agreeing on a hash for `SignedTransaction`,
                     // but both should be addressed.
                     let span =  info_span!(
                         "submit bundle",
                         nonce.initial = nonce,
                         bundle.len = bundle.len(),
+                        bundle.size = bundle.curr_size,
                     );
                     submission_fut = SubmitFut {
                         client: self.sequencer_client.clone(),
                         address: self.address,
                         nonce,
                         signing_key: self.sequencer_key.clone(),
                         state: SubmitState::NotStarted,
-                        bundle,
+                        bundle: bundle.into_actions(),
                     }
                     .instrument(span)
                     .fuse();
+                    }
+
+                // receive new seq_action and bundle it
+                Some(seq_action) = self.seq_actions_rx.recv(), if submission_fut.is_terminated() => {
+                    bundler.push(seq_action);
+                }
+
+                // try to preempt current bundle if the timer has ticked without submitting the next bundle
+                _ = block_timer.tick(), if submission_fut.is_terminated() => {
```

Comment:
The way it's written right now this can fire in a hot loop if the interval accumulated too long.